Add version query and integration tests

ashishajr opened the pull request:
# Description
Add `version` query
Add integration tests for `version` query for ok and failed

# Related Issue
[HTTP version explicit assert](https://github.com/Orange-OpenSource/hurl/issues/1706)

@jcamiel, I saw that in `assert_http_3_version.sh` it says that `HTTP/3` test doesn't pass in Arch Linux , I added the same to `assert_version.sh`. Let me know if I have to keep it or remove it 

```bash
set +eo pipefail
curl --version | grep Features | grep -q HTTP3
if [ $? -eq 1 ]; then
  exit 255
fi

# On désactive ce test momentanément Sur Arch Linux. Arch Linux supporte
# maintenant HTTP/3 mais ce test ne passe pas.
if [ -f /etc/os-release ]; then
    if grep -q 'NAME="Arch Linux"' /etc/os-release; then
        exit 255
    fi
fi
set -Eeuo pipefail
```

jcamiel commented on packages/hurl/src/runner/query.rs:
```diff
@@ -76,6 +76,19 @@ fn eval_query_status(response: &http::Response) -> QueryResult {
+        response
+            .version
+            .to_string()
+            .split("/")
```
Maybe instead of splitting, we can remove prefix `HTTP/` [strip_prefix](https://doc.rust-lang.org/std/primitive.str.html#method.strip_prefix)

ashishajr replied:
I've changed the query to use `strip_prefix` but I wanted to know why it's preferable to use `strip_prefix` instead of `split('/').last()`,

jcamiel replied:
`split` creates a iterator, and maybe a collection etc... I was thinking `strip_prefix` is sufficient here (even if `split('/').last()` is optimized not to create a collection)

jcamiel reviewed (changes requested):
Hi @ashishajr 

Thanks for the PR. Instead of adding new integration tests, I think we can modify existing ones that deals with HTTP versions (tehre are a lot of integration tests actually, we try to group them by uses-cases):

- `integration/hurl/tests_ok/http_version_2_option.sh`
- `integration/hurl/tests_ok/http_version_3_option.sh`
- `integration/hurl/tests_ok/http_version_10.sh`
- `integration/hurl/tests_ok/http_version_11.sh`
- `integration/hurl/tests_ok/http_version_option.sh`

For every implicit version, we can add also an explicit check by duplicating the request. For instance, with `integration/hurl/tests_ok/http_version_2_option.sh`:

Actual `integration/hurl/tests_ok/http_version_2_option.sh`, only implicit:

```hurl
# Live tests on google.com
HEAD https://google.com
HTTP/2 *


HEAD https://google.com
[Options]
http2: true
HTTP/2 *


HEAD https://google.com
[Options]
http2: false
HTTP/1.1 *


HEAD https://google.com
HTTP/2 *


HEAD https://google.com
HTTP/2 *
```

Updated `integration/hurl/tests_ok/http_version_2_option.sh` with additional explicit checks:

```hurl
# Implicit and explicit checks on HTTP version

# Live tests on google.com
HEAD https://google.com
HTTP/2 *


HEAD https://google.com
HTTP *
[Asserts]
version == "2"


HEAD https://google.com
[Options]
http2: true
HTTP/2 *


HEAD https://google.com
[Options]
http2: true
HTTP *
version == "2"


HEAD https://google.com
[Options]
http2: false
HTTP/1.1 *


HEAD https://google.com
[Options]
http2: false
HTTP *
[Asserts]
version == "1.1"


HEAD https://google.com
HTTP/2 *


HEAD https://google.com
HTTP *
[Asserts]
version = "2"


HEAD https://google.com
HTTP/2 *


HEAD https://google.com
HTTP *
[Asserts]
version = "2"
```

For the failed test `integration/hurl/tests_failed/assert_http_version.sh` we can do the same and add explicit version check.

Updated  `integration/hurl/tests_failed/assert_http_version.sh`:

```hurl
GET http://localhost:8000/error-assert/http-version
HTTP/2 200

GET http://localhost:8000/error-assert/http-version
HTTP 200
[Asserts]
version == "2"
```

In this case, we need to modify also `integration/hurl/tests_failed/assert_http_version.sh` and `integration/hurl/tests_failed/assert_http_version.ps1` so that we don't stop at the first error by adding a `--continue-on-error`:

Updated  `integration/hurl/tests_failed/assert_http_version.sh`: 

```bash
#!/bin/bash
set -Eeuo pipefail

hurl --continue-on-error tests_failed/assert_http_version.hurl
```

ashishajr commented:
Thanks a lot for the feedback @jcamiel, let me know if I missed something

ashishajr commented on integration/hurl/tests_ok/http_version_3_option.hurl:
```diff
@@ -27,3 +27,43 @@ HEAD https://google.com
 HTTP/1.1 *
+
+
+HEAD https://google.com
```
@jcamiel I had to move all these tests to the end instead of explicitly one after the other, for some reason when I did it by having the explicit assert after the implicit assert it would fail but doing it this way worked, I saw that the commit message makes a reference to force restarting the libcurl connection, could you please explain why it worked one way and not the other

jcamiel commented on integration/hurl/tests_ok/http_version_option.hurl:
```diff
@@ -13,34 +22,94 @@ GET http://localhost:8000/http_version/10
+GET http://localhost:8000/http_version/10
+[Options]
+http1.0: true
+HTTP/1.1 200
```
`HTTP/1.1 200` should be `HTTP 200` (we move the implicit check on an explicit ones)

jcamiel commented on integration/hurl/tests_ok/http_version_option.hurl:
```diff
@@ -3,6 +3,15 @@ GET http://localhost:8000/http_version/10
 HTTP 200
 
+
+GET http://localhost:8000/http_version/10
```
We can remove this request, in the previous request there is no implicit check (`HTTP 200`) so no need to add an explicit one

jcamiel commented on integration/hurl/tests_ok/http_version_option.hurl:
```diff
@@ -13,34 +22,94 @@ GET http://localhost:8000/http_version/10
 HTTP 200
 
+
+GET http://localhost:8000/http_version/11
```
We can remove this request, in the previous request we don't have an implicit version check (just `HTTP 200`) so no need to add an explicit check here.

jcamiel commented on integration/hurl/tests_ok/http_version_option.hurl:
```diff
@@ -13,34 +22,94 @@ GET http://localhost:8000/http_version/10
 HTTP 200
 
+
+GET http://localhost:8000/http_version/11
```
We can remove this request, in the previous request we don't have an implicit version check (just `HTTP 200`) so no need to add an explicit check here.

jcamiel commented on integration/hurl/tests_ok/http_version_option.hurl:
```diff
@@ -13,34 +22,94 @@ GET http://localhost:8000/http_version/10
 HTTP 200
 
+
+GET http://localhost:8000/http_version/11
```
We can remove this request, in the previous request we don't have an implicit version check (just `HTTP 200`) so no need to add an explicit check here.

jcamiel commented on integration/hurl/tests_ok/http_version_option.hurl:
```diff
@@ -13,34 +22,94 @@ GET http://localhost:8000/http_version/10
 HTTP 200
 
+
+GET http://localhost:8000/http_version/11
```
We can remove this request, in the previous request we don't have an implicit version check (just `HTTP 200`) so no need to add an explicit check here.

jcamiel commented on integration/hurl/tests_ok/http_version_option.hurl:
```diff
@@ -13,34 +22,94 @@ GET http://localhost:8000/http_version/10
+HEAD https://google.com
+[Options]
+http1.0: true
+HTTP/1.0 *
```
`HTTP/1.0 *`should be `HTTP *`(the implicit check becomes explicit)

jcamiel commented on integration/hurl/tests_ok/http_version_option.hurl:
```diff
@@ -13,34 +22,94 @@ GET http://localhost:8000/http_version/10
 [Options]
 http1.1: true
 HTTP/1.1 *
+[Asserts]
```
Same remarks, the implicit `HTTP/1.1 *` becomes `HTTP *`

jcamiel reviewed (changes requested):
Hi @ashishajr thanks for the PR some things to change. Just for an easier review, can you:

- when there is a version check (line `HTTP/1.1 *`): duplicate the request and transform it to an explicit check. Bu doing this, we should also transforms `HTTP/1.1 *`to `HTTP *`)

For instance this snippet:

```
HEAD https://google.com
[Options]
http1.0: true
HTTP/1.0 *
```

should becomes:

```
HEAD https://google.com
[Options]
http1.0: true
HTTP/1.0 *


HEAD https://google.com
[Options]
http1.0: true
HTTP *
[Asserts]
version = "1.1"
```

- if there is no version check (`HTTP *`), don't add an explicit test

- don't change order of the test _even if the tests faield_: there is certainly a bug in the current code (not your PR) and I want to identify it. Once I identify the usecase, I will adapt the Hurl file so this PR can be merge and will create an issue so we can fix the underlying bug

Thanks for the changes, it's really for me to make the review easier!!

jcamiel commented:
Thanks for the PR @ashishajr I will look at it asap

jcamiel commented:
Thanks @https://github.com/ashishajr/hurl.git I will look into it, we can't reinit the connection on everything request so I need so see iwhy the logic to detect change is buggy

jcamiel commented:
I've reduced the bug to just this file, independent of the new version query:

```hurl
HEAD https://google.com
HTTP/2 *

HEAD https://google.com
[Options]
http2: false
HTTP/1.1 *

HEAD https://google.com
[Options]
http2: false
HTTP/1.1 *
```

As it is idenpendant this PR could you: 

- just keep only one integration `tests_ok/http_version_option.hurl`  and the failded in `test_failed`
- rebase your commit and squash them in only one commit

I will make a new issue with this sample repro step and see what's wrong.

ashishajr commented:
Hey @jcamiel, just wanted to make sure I got it right, I need to put all the version tests into a single `tests_ok/http_version_option.hurl` right? For each option I'd need to first test it implicitly and then explicitly?

Does something like this look good?

```
HEAD https://google.com
[Options]
http1.0: true
HTTP/1.0 *


HEAD https://google.com
[Options]
http1.0: true
HTTP *
[Asserts]
version == "1.0"

HEAD https://google.com
[Options]
http1.1: true
HTTP/1.1 *


HEAD https://google.com
[Options]
http1.1: true
HTTP *
[Asserts]
version == "1.1"

// Other versions similarly ...
```

And the failed test in `tests_failed/http_version_option.hurl`

```
HEAD https://google.com
HTTP/2 *

HEAD https://google.com
[Options]
http2: false
HTTP/1.1 *

HEAD https://google.com
[Options]
http2: false
HTTP/1.1 *
```

jcamiel commented:
Yes @ashishajr I would like something simpler like this so all tests are green and we can merge it. I will update the tests once I fix the bug you raise. We can go with it and see if all tests are passing

ashishajr commented:
Hey @jcamiel, I've removed the older tests and moved all version tests into `tests_ok/http_version_option.hurl` and also added a `tests_failed/http_version_option.hurl` test. Let me know if the changes look good to you.

The bug seems to happen only with the latest version of curl, which I have to do to test HTTP3 changes, how do I modify the tests so the integration tests run fine in the pipeline?

Thanks for raising the bug in my place, I went into the rabbit hole of trying to fix it so and forgot to raise the bug report.

jcamiel commented:
Sorry @ashishajr I wasn't clear enough about what I wanted, my bad.

So, to recap:

- squahes all commits so the PR has only one commit
- only 1 failed test: `tests_failed/assert_http_version.hurl`

```
GET http://localhost:8000/error-assert/http-version
HTTP/2 200

GET http://localhost:8000/error-assert/http-version
HTTP 200
[Asserts]
version == "2"
```

- only one OK test: `tests_ok/http_version_option.hurl` with a new `version`assertion, for instance:

```hurl
GET http://localhost:8000/http_version/10
[Options]
http1.0: true
HTTP 200

# RFC 2616 requires that HTTP servers always begin their responses with
# the highest HTTP version that they claim to support. So the HTTP version can
# be different from what the client used in their request.
# See <https://serverfault.com/questions/1059870/why-does-apache-send-http-1-1-when-client-requests-http-1-0-and-many-other-very>
# We validate on the Flask endpoint `http_version.py` that the version asked by the client is the right one.
GET http://localhost:8000/http_version/10
[Options]
http1.0: true
HTTP/1.1 200

GET http://localhost:8000/http_version/10
[Options]
http1.0: true
HTTP 200
[Asserts]
version == "1.1"


# Default HTTP version for Flask is HTTP/1.1
GET http://localhost:8000/http_version/11
HTTP 200

GET http://localhost:8000/http_version/11
[Options]
http1.1: true
HTTP 200

# The last option win
GET http://localhost:8000/http_version/11
[Options]
http1.0: true
http1.1: true
HTTP 200

GET http://localhost:8000/http_version/11
[Options]
http1.0: false
HTTP 200

# Live test on google.fr
HEAD https://google.com
[Options]
http1.0: true
HTTP/1.0 *

HEAD https://google.com
[Options]
http1.1: true
HTTP/1.1 *
```

I want to have as few as diff as possible in this PR. I will update the tests once the issue #3719 is resolved


Many thanks and sorry for the changes!

ashishajr commented:
Thanks for the clarification @jcamiel, I misunderstood the instructions a bit 😅

jcamiel commented on integration/hurl/tests_ok/http_version_10.ps1:
This file musn't be touch nor deleted

jcamiel commented on integration/hurl/tests_ok/http_version_10.sh:
This file musn't be touch nor deleted

jcamiel commented on integration/hurl/tests_ok/http_version_11.hurl:
This file musn't be touch nor deleted

jcamiel commented on integration/hurl/tests_ok/http_version_11.curl:
This file musn't be touch nor deleted

jcamiel commented on integration/hurl/tests_ok/http_version_11.ps1:
This file musn't be touch nor deleted

jcamiel commented on integration/hurl/tests_ok/http_version_11.sh:
This file musn't be touch nor deleted

jcamiel commented on integration/hurl/tests_ok/http_version_2_option.hurl:
This file musn't be touch nor deleted

jcamiel commented on integration/hurl/tests_ok/http_version_2_option.out:
This file musn't be touch nor deleted

jcamiel commented on integration/hurl/tests_ok/http_version_2_option.sh:
This file musn't be touch nor deleted

jcamiel commented on integration/hurl/tests_ok/http_version_3_option.hurl:
This file musn't be touch nor deleted

jcamiel commented on integration/hurl/tests_ok/http_version_3_option.ps1:
This file musn't be touch nor deleted

jcamiel commented on integration/hurl/tests_ok/http_version_3_option.sh:
This file musn't be touch nor deleted

jcamiel commented:
/accept
